### PR TITLE
More cleanups around trivial input forwarding and aliasing.

### PR DESCRIPTION
**Pull request by wujingyue**
This PR removed several unnecessary special cases for trivial input forwarding and aliasing. In most cases, e.g. when computing tensor proxies, forwarded outputs and aliased outputs can simply be treated as regular outputs. 

Similar to #1141 , I verified the number of memory allocation requests for trivial input forwarding. 

```
$ PYTORCH_NO_CUDA_MEMORY_CACHING=1 nsys nvprof bin/nvfuser_tests --gtest_filter=*TrivialInputForwarding_FusionExecutor 2> /dev/null | grep cudaMalloc
      1.1            89400          2    44700.0    44700.0      3607     85793      58114.3  cudaMalloc
```

### CODE REVIEW

File: test/test_gpu3.cpp [hunk 1]
```diff
@@ -5227,7 +5227,32 @@ TEST_F(NVFuserTest, FusionInlineAt_CUDA) {
...
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor t0 = at::randn({10, 4}, options);
+  at::Tensor t1 = at::randn({10, 4}, options);
+
+  FusionExecutor fe;
+  fe.compileFusion(&fusion, {t0, t1});
+  at::Tensor t0_forward = fe.runFusion({t0, t1})[0];
+
+  EXPECT_EQ(t0_forward.data_ptr(), t0.data_ptr());
```
Comment (wujingyue) [inline test/test_gpu3.cpp:5251]: TODO: this check used to fail without this PR, but the same check in the FusionExecutorCache version was fine. This suggests FusionExecutorCache might be doing some output handling that's unnecessary now.

File: csrc/executor.cpp [hunk 2]
```diff
@@ -1236,37 +1238,19 @@ std::vector<FusionExecutor::GlobalBufferInfo> FusionExecutor::
...
-      NVF_ERROR(
-          info.tv != nullptr, "Cannot allocate outputs that are not tensors.");
-      auto output = out_val->as<TensorView>();
-      auto alias_it = std::find_if(
-          output_to_input_aliases.begin(),
-          output_to_input_aliases.end(),
-          [&](const auto output_to_input) {
-            return output_to_input.first == (int)out_i;
-          });
-      if (alias_it != output_to_input_aliases.end()) {
```
Comment (wujingyue) [inline csrc/executor.cpp:1257]: TODO: remove the `output_to_input_aliases` argument.

**Comment by wujingyue** [comment]
!build

**Comment by wujingyue** [comment]
!build

**Comment by wujingyue** [comment]
The build error looks unrelated to the PR: 

```
FAILED: lib/dynamic_type/CMakeFiles/test_dynamic_type_23.dir/test/ForAllTypes.cpp.o 
/usr/bin/clang++  -I/home/runner/work/Fuser/Fuser/cmake/../third_party/benchmark/include -I/home/runner/work/Fuser/Fuser/lib/dynamic_type/src -isystem /home/runner/work/Fuser/Fuser/cmake/../third_party/googletest/googlemock/include -isystem /home/runner/work/Fuser/Fuser/cmake/../third_party/googletest/googletest/include -isystem /home/runner/work/Fuser/Fuser/cmake/../third_party/flatbuffers/include -isystem /home/runner/work/Fuser/Fuser/third_party/googletest/googletest/include -isystem /home/runner/work/Fuser/Fuser/third_party/googletest/googletest -isystem /home/runner/work/Fuser/Fuser/third_party/googletest/googlemock/include -isystem /home/runner/work/Fuser/Fuser/third_party/googletest/googlemock -Wno-psabi -D_GLIBCXX_USE_CXX11_ABI=0 -O3 -DNDEBUG -std=gnu++2b -MD -MT lib/dynamic_type/CMakeFiles/test_dynamic_type_23.dir/test/ForAllTypes.cpp.o -MF lib/dynamic_type/CMakeFiles/test_dynamic_type_23.dir/test/ForAllTypes.cpp.o.d -o lib/dynamic_type/CMakeFiles/test_dynamic_type_23.dir/test/ForAllTypes.cpp.o -c /home/runner/work/Fuser/Fuser/lib/dynamic_type/test/ForAllTypes.cpp
In file included from /home/runner/work/Fuser/Fuser/lib/dynamic_type/test/ForAllTypes.cpp:9:
In file included from /home/runner/work/Fuser/Fuser/lib/dynamic_type/src/dynamic_type/type_traits.h:10:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/tuple:44:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/ranges_util.h:34:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/ranges_base.h:37:
/usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/stl_iterator.h:2618:35: error: missing 'typename' prior to dependent type name 'iterator_traits<_It>::iterator_category'
      { using iterator_category = iterator_traits<_It>::iterator_category; };
                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

`-std=gnu++2b` suggests we were building with C++23 support?! I thought we didn't intend to support anything beyond C++20.

**Comment by zasdfgbnm** [comment]
> > The build error looks unrelated to the PR:
> > ```
> > FAILED: lib/dynamic_type/CMakeFiles/test_dynamic_type_23.dir/test/ForAllTypes.cpp.o 
> > /usr/bin/clang++  -I/home/runner/work/Fuser/Fuser/cmake/../third_party/benchmark/include -I/home/runner/work/Fuser/Fuser/lib/dynamic_type/src -isystem /home/runner/work/Fuser/Fuser/cmake/../third_party/googletest/googlemock/include -isystem /home/runner/work/Fuser/Fuser/cmake/../third_party/googletest/googletest/include -isystem /home/runner/work/Fuser/Fuser/cmake/../third_party/flatbuffers/include -isystem /home/runner/work/Fuser/Fuser/third_party/googletest/googletest/include -isystem /home/runner/work/Fuser/Fuser/third_party/googletest/googletest -isystem /home/runner/work/Fuser/Fuser/third_party/googletest/googlemock/include -isystem /home/runner/work/Fuser/Fuser/third_party/googletest/googlemock -Wno-psabi -D_GLIBCXX_USE_CXX11_ABI=0 -O3 -DNDEBUG -std=gnu++2b -MD -MT lib/dynamic_type/CMakeFiles/test_dynamic_type_23.dir/test/ForAllTypes.cpp.o -MF lib/dynamic_type/CMakeFiles/test_dynamic_type_23.dir/test/ForAllTypes.cpp.o.d -o lib/dynamic_type/CMakeFiles/test_dynamic_type_23.dir/test/ForAllTypes.cpp.o -c /home/runner/work/Fuser/Fuser/lib/dynamic_type/test/ForAllTypes.cpp
> > In file included from /home/runner/work/Fuser/Fuser/lib/dynamic_type/test/ForAllTypes.cpp:9:
> > In file included from /home/runner/work/Fuser/Fuser/lib/dynamic_type/src/dynamic_type/type_traits.h:10:
> > In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/tuple:44:
> > In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/ranges_util.h:34:
> > In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/ranges_base.h:37:
> > /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/stl_iterator.h:2618:35: error: missing 'typename' prior to dependent type name 'iterator_traits<_It>::iterator_category'
> >       { using iterator_category = iterator_traits<_It>::iterator_category; };
> >                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> > 1 error generated.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `-std=gnu++2b` suggests we were building with C++23 support?! I thought we didn't intend to support anything beyond C++20.
> 
> cc @xwang233 and @zasdfgbnm

We do. The main purpose for building it with C++23 is to make sure we do not use any feature that will be deprecated in C++23. But if you think what you hit is just a compiler bug, I don't mind just disabling it. There is no reason to workaround compiler bugs for better support C++23.

**Comment by zasdfgbnm** [comment]
You might want to comment out this line: https://github.com/NVIDIA/Fuser/blob/a4f266e28e4689bec3a87135c89ce7a742ea6f54/lib/dynamic_type/CMakeLists.txt#L52

**Comment by wujingyue** [comment]
> We do. The main purpose for building it with C++23 is to make sure we do not use any feature that will be deprecated in C++23. But if you think what you hit is just a compiler bug, I don't mind just disabling it. There is no reason to workaround compiler bugs for better support C++23.

That's fair. @zasdfgbnm , can you take a look at this build error? It was triggered by `lib/dynamic_type/test/ForAllTypes.cpp` and appeared in some STL header file.

File: test/test_gpu3.cpp [hunk 3]
```diff
@@ -5227,7 +5227,31 @@ TEST_F(NVFuserTest, FusionInlineAt_CUDA) {
...
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  TensorView* tv0 = makeConcreteTensor({-1, -1});
+  TensorView* tv1 = makeConcreteTensor({-1, -1});
+  fusion.addInput(tv0);
+  fusion.addInput(tv1);
+  // Note: output of add is not used. Kept it here since previously there was an
+  // assertion from sorting in codegen.
+  add(tv1, IrBuilder::create<Val>(3.141));
```
Comment (jjsjann123) [inline test/test_gpu3.cpp:5240]: nitpick: is this still the case that we needed the add here?

**Comment by zasdfgbnm** [comment]
> > We do. The main purpose for building it with C++23 is to make sure we do not use any feature that will be deprecated in C++23. But if you think what you hit is just a compiler bug, I don't mind just disabling it. There is no reason to workaround compiler bugs for better support C++23.
> 
> That's fair. @zasdfgbnm , can you take a look at this build error? It was triggered by `lib/dynamic_type/test/ForAllTypes.cpp` and appeared in some STL header file.

It is triggered at `ForAllTypes.cpp:9`, which is just
https://github.com/NVIDIA/Fuser/blob/a4f266e28e4689bec3a87135c89ce7a742ea6f54/lib/dynamic_type/test/ForAllTypes.cpp#L9
where in `type_traits.h`, the failing line is:
https://github.com/NVIDIA/Fuser/blob/a4f266e28e4689bec3a87135c89ce7a742ea6f54/lib/dynamic_type/src/dynamic_type/type_traits.h#L10
So a minimum repro for this compilation error should be as simple as
```C++
#include <tuple>
```
I am 90% sure that this is either a bug in compiler or a bug in the standard library. Whichever it is, let's just disable the C++23 build.

File: csrc/kernel_cache.cpp [hunk 4]
```diff
@@ -1216,61 +1220,10 @@ std::vector<at::Tensor> FusionKernelRuntime::runWithInputs(
...
-      // the code. However, it is required to be a kernel argument, which acts
-      // as a place-holder, so we can map the arguments correctly.
-      //
-      // 2) Integration handles the trivial forwarding of inputs. When we put
-      // together `fusion_outputs` for a given fusion and the outputs are
-      // fusion inputs, we directly return the input tensor.
-      NVF_ERROR(iter->second->is<at::Tensor>());
-      fusion_outputs.push_back(iter->second->as<at::Tensor>());
-    } else {
-      bool empty_type_check = output->getDataType().has_value() &&
```
Comment (jjsjann123) [inline csrc/kernel_cache.cpp:1245]: Nice clean up. Wasn't aware that these are dead code now. :clap:

**Review by jjsjann123 — APPROVED** [review]
LGTM

File: test/test_gpu3.cpp (hunk 3 above)
Comment (wujingyue) [reply test/test_gpu3.cpp]: It's good to leave it as a regression test. But since it's already in another test, let's remove it.

**Pull request merged**